pkg/controller/direct/monitoring/utils.go

```diff
@@ -25,3 +42,142 @@ func ValueOf[T any](p *T) T {
 func PtrTo[T any](t T) *T {
 	return &t
 }
+
+// IsNotFound returns true if the given error is an HTTP 404.
+func IsNotFound(err error) bool {
+	return HasHTTPCode(err, 404)
+}
+
+// HasHTTPCode returns true if the given error is an HTTP response with the given code.
+func HasHTTPCode(err error, code int) bool {
+	if err == nil {
+		return false
+	}
+	apiError := &apierror.APIError{}
+	if errors.As(err, &apiError) {
+		if apiError.HTTPCode() == code {
+			return true
+		}
+	} else {
+		klog.Warningf("unexpected error type %T", err)
+	}
+	return false
+}
+
+func lastComponent(s string) string {
+	i := strings.LastIndex(s, "/")
+	return s[i+1:]
+}
+
+func ComputeChangedFields(actual proto.Message, desired proto.Message) sets.Set[string] {
+	changes := sets.New[string]()
+	actualReflect := actual.ProtoReflect()
+	desiredReflect := desired.ProtoReflect()
+	actualReflect.Range(func(field protoreflect.FieldDescriptor, actualValue protoreflect.Value) bool {
+		desiredValue := desiredReflect.Get(field)
+		if !actualValue.Equal(desiredValue) {
+			changes.Insert(field.JSONName())
+		}
+		return true
+	})
+	desiredReflect.Range(func(field protoreflect.FieldDescriptor, desiredValue protoreflect.Value) bool {
+		actualValue := actualReflect.Get(field)
+		if !actualValue.Equal(desiredValue) {
+			changes.Insert(field.JSONName())
+		}
+		return true
+	})
+	if changes.Len() != 0 {
+		klog.V(2).Infof("ComputeChangedFields found diff fields=%v, diff=%v", sets.List(changes), cmp.Diff(actual, desired, protocmp.Transform()))
+	}
+	return changes
+}
+
+func setStatus(u *unstructured.Unstructured, typedStatus any) error {
+	status, err := runtime.DefaultUnstructuredConverter.ToUnstructured(typedStatus)
+	if err != nil {
+		return fmt.Errorf("error converting status to unstructured: %w", err)
+	}
+
+	// Use existing values for conditions/observedGeneration; they are managed in k8s not the GCP API
+	old, _, _ := unstructured.NestedMap(u.Object, "status")
+	if old != nil {
+		status["conditions"] = old["conditions"]
+		status["observedGeneration"] = old["observedGeneration"]
+	}
+
+	u.Object["status"] = status
+
+	return nil
+}
+
```

Comment:
qq: do you think we shall move this to the direct_controller`Reconcile`?

Reply:
We should move it, along with most/all of the functions in utils.  I just don't think we know where to move it to yet!